File: src/main/java/org/cloudio/morpheus/mail/AlternatingUserMail.java
package org.cloudio.morpheus.mail;

/**
 * Created by zslajchrt on 26/08/15.
 */
public abstract class AlternatingUserMail implements UserMail, FaxByMail {

    protected abstract UserMail getDelegate();

    public void sendEmail(Message message) {
        getDelegate().sendEmail(message);
    }

    public void validateEmail(Message message) {
        getDelegate().validateEmail(message);
    }

    public boolean canFaxEmail() {
        return getDelegate() instanceof FaxByMail;
    }

    public void faxEmail(Message message) {
        UserMail delegate = getDelegate();
        if (delegate instanceof FaxByMail) {
            ((FaxByMail) delegate).faxEmail(message);
        } else {
            throw new IllegalStateException("The current service does not support fax");
        }
    }
}

File: src/main/scala/org/cloudio/morpheus/mail/traditional/AlternatingUserMail.scala
package org.cloudio.morpheus.mail.traditional


/**
 * Created by zslajchrt on 26/08/15.
 */
abstract class AlternatingUserMail extends UserMail {

  override def sendEmail(message: Message) {
    getDelegate.sendEmail(message)
  }

  override def validateEmail(message: Message) {
    getDelegate.validateEmail(message)
  }

  protected def getDelegate: UserMail

  def canFaxEmail: Boolean = getDelegate.isInstanceOf[FaxByMail]

  def faxEmail(message: Message) {
    getDelegate match {
      case fax: FaxByMail =>
        fax.faxEmail(message)
      case _ =>
        throw new IllegalStateException("The current service does not support fax")
    }
  }
}
